examples/motoko_examples/basic-dao/src/basic_dao.ts

```diff
@@ -0,0 +1,374 @@
+  err?: string;
+}>;
+// Vote on an open proposal
+export function vote(args: VoteArgs): Update<VoteResult> {
+  let caller = ic.caller();
+  let proposal = proposalGet(args.proposal_id);
+  if (proposal === null) {
+    return { err: `No proposal with ID  ${args.proposal_id} exists` };
+  }
+  // I am keeping the if statemnts here even though they are not needed for readability
+  // as this whole block should only be excuted if there's a proposal
+  if (proposal) {
+    let state = proposal.state;
+    if (!('Open' in state)) {
+      return {
+        err: `Proposal ${args.proposal_id} is not open for voting"`,
+      };
+    }
+    let account = accountGet(caller);
+    if (account === null) {
+      return { err: 'Caller does not have any tokens to vote with' };
+    }
+    if (account) {
+      if (proposal.voters.some((e) => e === caller)) {
+        return { err: 'Already voted' };
+      }
+
+      let votes_yes = proposal.votes_yes.amount_e8s;
+      let votes_no = proposal.votes_no.amount_e8s;
+      if ('Yes' in args.vote) {
+        votes_yes += account.amount_e8s;
+      }
+      if ('No' in args.vote) {
+        votes_no += account.amount_e8s;
+      }
+      proposal.voters.push(caller);
+      const voters = proposal.voters;
+
+      if (
+        votes_yes >=
+        ic.stableStorage<StableStorage>().systemParams.proposal_vote_threshold
+          .amount_e8s
+      ) {
+        // Refund the proposal deposit when the proposal is accepted
+        const account = accountGet(proposal.proposer)!;
+        const refunded =
+          account.amount_e8s +
+          ic.stableStorage<StableStorage>().systemParams
+            .proposal_submission_deposit.amount_e8s;
+        account_put(proposal.proposer, { amount_e8s: refunded });
+
+        state = { Accepted: null };
+      }
+
+      if (
+        votes_no >=
+        ic.stableStorage<StableStorage>().systemParams.proposal_vote_threshold
+          .amount_e8s
+      ) {
+        state = { Rejected: null };
+      }
+
+      const updated_proposal = {
+        id: proposal.id,
+        votes_yes: { amount_e8s: votes_yes },
+        votes_no: { amount_e8s: votes_no },
+        voters,
+        state,
+        timestamp: proposal.timestamp,
+        proposer: proposal.proposer,
+        payload: proposal.payload,
+      };
+      proposalPut(args.proposal_id, updated_proposal);
+    }
+
+    return { ok: state };
+  }
+
+  return { err: 'Unexpected error' };
+}
+
+/// Get the current system params
+export function getSystemParams(): Query<SystemParams> {
+  return ic.stableStorage<StableStorage>().systemParams;
+}
+
+/// Update system params
+///
+/// Only callable via proposal execution
+export function updateSystemParams(
+  payload: UpdateSystemParamsPayload
+): Update<void> {
+  let caller = ic.caller();
+  if (caller != ic.id()) {
+    return;
+  }
+  ic.stableStorage<StableStorage>().systemParams = {
+    ...ic.stableStorage<StableStorage>().systemParams,
```

Comment:
This will probably need to be reworked once `UpdateSystemParamsPayload` uses optional types.